primitives/src/swap/swap.rs

```diff
@@ -111,144 +111,131 @@ pub struct Swap<AccountId, BlockNumber> {
-        )?),
-    )
+      );
+    }
+
+    Err(SlippageError::NoUpperBoundForSelling)
   }
 
   fn pay_per_token<FT, FF>(
     &self,
-    amount_to_closure: FT,
-    amount_from_closure: FF,
+    base_amount_closure: FT,
+    quote_amount_closure: FF,
+    market_pair: &MarketPair<CurrencyId>,
+    base_amount: Balance,
+    quote_amount: Balance,
   ) -> Result<FixedU128, SlippageError>
   where
     FT: Fn(Balance) -> Balance,
     FF: Fn(Balance) -> Balance,
   {
-    let token_to: Asset = self
-      .token_to
+    let base_asset: Asset = market_pair.base_asset
       .try_into()
       .map_err(|_| SlippageError::UnknownAsset)?;
 
-    let token_to_one_unit = token_to.saturating_mul(1);
+    let base_token_one_unit = base_asset.saturating_mul(1);
 
-    let token_from: Asset = self
-      .token_from
+    let quote_asset: Asset = market_pair.quote_asset
       .try_into()
       .map_err(|_| SlippageError::UnknownAsset)?;
-    let token_from_one_unit = token_from.saturating_mul(1);
+    let quote_token_one_unit = quote_asset.saturating_mul(1);
 
-    FixedU128::saturating_from_rational(amount_to_closure(self.amount_to), token_to_one_unit)
+    FixedU128::saturating_from_rational(quote_amount_closure(quote_amount), quote_token_one_unit)
       .checked_div(&FixedU128::saturating_from_rational(
-        amount_from_closure(self.amount_from),
-        token_from_one_unit,
+        base_amount_closure(base_amount),
+        base_token_one_unit
       ))
       .ok_or(SlippageError::SlippageOverflow)
   }
 
   fn validate_slippage_dry_run(
     &self,
-    amount_to_receive: Balance,
-    amount_to_send: Balance,
+    price_offered: FixedU128,
+    market_pair: &MarketPair<CurrencyId>,
   ) -> Result<(), SlippageError> {
-    let token_to: Asset = self
-      .token_to
-      .try_into()
-      .map_err(|_| SlippageError::UnknownAsset)?;
-
-    let token_to_one_unit = token_to.saturating_mul(1);
-
-    let token_from: Asset = self
-      .token_from
-      .try_into()
-      .map_err(|_| SlippageError::UnknownAsset)?;
-    let token_from_one_unit = token_from.saturating_mul(1);
-
-    let pay_per_token_lower_bond = self.pay_per_token_lower_bond()?;
-    let pay_per_token_upper_bond = self.pay_per_token_upper_bond()?;
-
-    let pay_per_token_offered =
-      FixedU128::saturating_from_rational(amount_to_receive, token_to_one_unit)
-        .checked_div(&FixedU128::saturating_from_rational(
-          amount_to_send,
-          token_from_one_unit,
-        ))
-        .ok_or(SlippageError::SlippageOverflow)?;
-
-    // limit order can match with smaller price
-    if self.swap_type != SwapType::Limit {
-      if pay_per_token_offered.lt(&pay_per_token_lower_bond) {
+    if self.token_to == market_pair.base_asset { // buyer should not accept a price greater than upper bound
+      let pay_per_token_upper_bond = self.pay_per_token_upper_bond(market_pair)?;
+      if price_offered.gt(&pay_per_token_upper_bond) {
+        if self.is_market_maker {
```

Comment:
Being a market maker has no effect in slippage. You can be a market maker and do market swaps instead of limit swaps

Reply:
Yes, this condition is just for specifying the error is from the swap created by trader or market maker.